sui-execution/latest/sui-move-natives/src/transfer.rs

```diff
@@ -133,14 +156,25 @@ pub fn share_object(
     let cost = context.gas_used();
     Ok(match transfer_result {
         // New means the ID was created in this transaction
-        // SameOwner means the object was previously shared and was re-shared; since
-        // shared objects cannot be taken by-value in the adapter, this can only
-        // happen via test_scenario
+        // SameOwner means the object was previously shared and was re-shared
         TransferResult::New | TransferResult::SameOwner => NativeResult::ok(cost, smallvec![]),
         TransferResult::OwnerChanged => NativeResult::err(cost, E_SHARED_NON_NEW_OBJECT),
     })
 }
 
+pub fn object_is_shared(context: &mut NativeContext, obj: &Value) -> PartialVMResult<bool> {
```

Comment:
It feels a bit strange to me that this is a separate function rather than being a part of `object_runtime_transfer`

Thoughts on slapping a `#[must_use]` on `TransferResult` into `ObjectRuntime::transfer`? 
The function has to already load and inspect the input state, and now we are saying that some combination of old+new owner states are an error. So it makes sense to me that `TransferResult` would just carry that information

Reply:
I like the idea of passing the Transfer result into the `object_runtime_transfer` function to do the checks there, however the logic is different for all the functions so not sure how to pull this off neatly. What do you mean by must_use?

Reply:
I'm suggesting changing the logic of `ObjectRuntime::transfer` so that `TransferResult` has a case for this error. Something like 
```
TransferResult::InvalidTransfer(/* reason of some sort? not sure what info you need */)
```

Putting `#[must_use]` on `TransferResult` will ensure that call-sites consider the value, as you normally do with `Result` ([further reading](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute))